lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,241 @@
+          post[:cardOnFile][:indicator] = options[:card_on_file][:transaction_id]
+        end
+        if post[:hotel].present?
+          post[:hotel] = {}
+          post[:hotel][:estimatedDays] = options[:hotel][:estimated_days]
+        end
+        if post[:pin].present?
+          post[:pin] = {}
+          post[:pin][:block] = options[:pin][:block]
+          post[:pin][:ksn] = options[:pin][:ksn]
+        end
+      end
+
+      def add_device(post, options)
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, options)
+        post[:card] = {}
+        if options[:card][:token].present?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = options[:card][:token][:value]
+        else
+          post[:card][:expirationDate] = options[:card][:expiration_date]
+          post[:card][:number] = options[:card][:number]
+          post[:card][:entryMode] = options[:card][:entry_mode]
+          post[:card][:present] = options[:card][:present]
+          post[:card][:securityCode] = {}
+          if options[:card][:security_code].present?
+            post[:card][:securityCode][:indicator] = options[:card][:security_code][:indicator]
+            post[:card][:securityCode][:value] = options[:card][:security_code][:value]
+          end
+        end
+      end
+
+      def add_customer(post, options)
+        if options[:customer].present?
+          post[:customer] = {}
+          post[:customer][:addressLine1] = options[:customer][:address_line1]
+          post[:customer][:firstName] = options[:customer][:first_name]
+          post[:customer][:lastName] = options[:customer][:last_name]
+          post[:customer][:postalCode] = options[:customer][:postal_code]
+        end
+      end
+
+      def commit(action, parameters, headers)
+        url_postfix = action == 'accesstoken' ? 'credentials' : 'transactions'
+        url = (test? ? "#{test_url}#{url_postfix}/#{action}" : "#{live_url}#{url_postfix}/#{action}")
+        url = URI(url)
+        https = Net::HTTP.new(url.host, url.port)
```

Comment:
We should be able to use the existing `ssl_post` and `ssl_request` methods here: https://github.com/activemerchant/active_merchant/blob/a9c588f90c4c1d59f7b0f5e1dc40b9ca12cb7fbf/lib/active_merchant/posts_data.rb#L39

Reply:
resolved.